hpcbench/benchmark/custream.py

```diff
@@ -0,0 +1,176 @@
+        copy_bandwidth=Metrics.GigaBytesPerSecond,
+        scale_min_time=Metrics.Second,
+        scale_avg_time=Metrics.Second,
+        scale_max_time=Metrics.Second,
+        scale_bandwidth=Metrics.GigaBytesPerSecond,
+        add_min_time=Metrics.Second,
+        add_avg_time=Metrics.Second,
+        add_max_time=Metrics.Second,
+        add_bandwidth=Metrics.GigaBytesPerSecond,
+        triad_min_time=Metrics.Second,
+        triad_avg_time=Metrics.Second,
+        triad_max_time=Metrics.Second,
+        triad_bandwidth=Metrics.GigaBytesPerSecond,
+    )
+
+    METRICS_NAMES = set(METRICS)
+
+    @property
+    def metrics(self):
+        """ The metrics to be extracted.
+            This property can not be replaced, but can be mutated as required
+        """
+        return CUDAStreamExtractor.METRICS
+
+    def extract_metrics(self, outdir, metas):
+        metrics = {}
+        # parse stdout and extract desired metrics
+        with open(self.stdout(outdir)) as istr:
+            for line in istr:
+                if line.strip() in self.STDOUT_IGNORE_PRIOR:
+                    break
+            for line in istr:
+                line = line.strip()
+                CUDAStreamExtractor._parse_line(line, metrics)
+        return metrics
+
+    @classmethod
+    def _parse_line(cls, line, metrics):
+        for sect, regex in cls.REGEX.items():
+            search = regex.search(line)
+            if search:
+                metrics[sect + "_bandwidth"] = float(search.group(1))
+                metrics[sect + "_avg_time"] = float(search.group(2))
+                metrics[sect + "_min_time"] = float(search.group(3))
+                metrics[sect + "_max_time"] = float(search.group(4))
+                return
+
+
+class CUDAStream(Benchmark):
+    """Benchmark wrapper for the cuda-stream utility
+    """
+    name = 'custream'
+
+    description = "Provides memory bandwidth benchmarking for NVIDIA GPUs."
+
+    DEFAULT_EXECUTABLE = 'stream'
+    DEFAULT_THREADS_PER_BLOCK = [64, 128, 192, 256, 512, 1024]
+
+    def __init__(self):
+        super(CUDAStream, self).__init__(
+            attributes=dict(
+                blocksizes=CUDAStream.DEFAULT_THREADS_PER_BLOCK,
+                executable=CUDAStream.DEFAULT_EXECUTABLE,
+            )
+        )
+
+    @cached_property
+    def executable(self):
+        """Get path to iperf executable
+        """
+        return self.attributes['executable']
+
+    @property
+    def blocksizes(self):
+        """List of threads per block the command is executed with"""
+        return [
+            str(e)
+            for e in self.attributes['blocksizes']
+        ]
+
+    def execution_matrix(self, context):
+        del context  # unused
+        for nthreads in self.blocksizes:
+            yield dict(
+                category=CUDAStream.name,
+                command=[find_executable(self.executable),
+                         '-b {threads}'.format(threads=nthreads)
```

Comment:
If `stream "-b 64"` commands works, it is fine, but I doubt it.
I think it should be: `[find_executable(self.executable), '-b', nthreads]`.

Reply:
you're obviously right - fixed.